File: ruby/protein-translation/protein_translation.rb
class InvalidCodonError < StandardError ; end

class Translation

  PROTEIN_CODONS = {
    "Methionine"    => ['AUG'],
    "Phenylalanine" => ['UUU', 'UUC'],
    "Leucine"       => ['UUA', 'UUG'],
    "Serine"        => ['UCU', 'UCC', 'UCA', 'UCG'],
    "Tyrosine"      => ['UAU', 'UAC'],
    "Cysteine"      => ['UGU', 'UGC'],
    "Tryptophan"    => ['UGG'],
    "STOP"          => ['UAA', 'UAG', 'UGA']
  }.freeze

  def self.of_codon(codon)
    protein = PROTEIN_CODONS.find { |k, v| v.include? codon }
    raise InvalidCodonError if protein.nil?
    protein[0]
  end

  def self.of_rna(rna)
    rna.scan(/\w{3}/)
       .map { |codon| Translation.of_codon(codon) }
       .take_while { |e| e != 'STOP' }
  end

end

File: python/protein-translation/protein_translation.py
PROTEIN_CODONS = {
    "Methionine": ['AUG'],
    "Phenylalanine": ['UUU', 'UUC'],
    "Leucine": ['UUA', 'UUG'],
    "Serine": ['UCU', 'UCC', 'UCA', 'UCG'],
    "Tyrosine": ['UAU', 'UAC'],
    "Cysteine": ['UGU', 'UGC'],
    "Tryptophan": ['UGG'],
    "STOP": ['UAA', 'UAG', 'UGA']
  }


def proteins(strand):
    codons = []
    strand_len = len(strand)
    for i in range(0,strand_len,3):
        codon = of_codon(strand[i:min(strand_len, i+3)])
        if codon == 'STOP':
            break
        codons.append(codon)
    return codons

def of_codon(strand):
    codon = ''
    for k, v in PROTEIN_CODONS.items():
        if strand in v:
            codon = k
            break
    return codon
